gui/src/app/SamplingOptsPanel/SamplingOptsPanel.tsx

```diff
@@ -0,0 +1,166 @@
+                    <tr>
+                        <td>init radius</td>
+                        <td>
+                            <FloatEdit
+                                value={samplingOpts.init_radius}
+                                onChange={(value) => setSamplingOpts && setSamplingOpts({ ...samplingOpts, init_radius: value })}
+                                min={0}
+                                readOnly={readOnly}
+                            />
+                        </td>
+                    </tr>
+                    <tr>
+                        <td>seed</td>
+                        <td>
+                            <IntOrUndefinedEdit
+                                value={samplingOpts.seed}
+                                onChange={(value) => setSamplingOpts && setSamplingOpts({ ...samplingOpts, seed: value })}
+                                min={0}
+                                readOnly={readOnly}
+                            />
+                        </td>
+                    </tr>
+                </tbody>
+            </table>
+            <div>
+                <Hyperlink onClick={handleReset} color="gray">reset</Hyperlink>
+            </div>
+        </div>
+    )
+}
+
+type InitEditProps = {
+    value: number;
+    onChange: (value: number) => void;
+    min: number;
+    max?: number;
+    readOnly: boolean;
+}
+
+const IntEdit: FunctionComponent<InitEditProps> = ({ value, onChange, min, max, readOnly }) => {
+    return (
+        <input
+            type="number"
+            value={value}
+            onChange={(e) => onChange(parseInt(e.target.value))}
+            min={min}
+            max={max}
+            readOnly={readOnly}
+            style={{ width: "4em" }}
+        />
+    )
+}
+
+type FloatEditProps = {
+    value: number;
+    onChange: (value: number) => void;
+    min: number;
+    max?: number;
+    readOnly: boolean;
+}
+
+const FloatEdit: FunctionComponent<FloatEditProps> = ({ value, onChange, min, max, readOnly }) => {
+    return (
+        <input
+            type="number"
+            value={value}
+            onChange={(e) => onChange(parseFloat(e.target.value))}
+            min={min}
+            max={max}
+            readOnly={readOnly}
+            style={{ width: "4em" }}
+        />
+    )
+}
+
+type IntOrUndefinedEditProps = {
+    value: number | undefined;
+    onChange: (value: number | undefined) => void;
+    min: number;
+    readOnly: boolean;
+}
+
+const IntOrUndefinedEdit: FunctionComponent<IntOrUndefinedEditProps> = ({ value, onChange, min, readOnly }) => {
+    return (
+        <input
+            type="number"
+            value={value === undefined ? "" : value}
+            onChange={(e) => {
+                const value = parseInt(e.target.value);
+                onChange(isNaN(value) ? undefined : value)
+            }}
+            min={min}
+            readOnly={readOnly}
+            style={{ width: "4em" }}
+        />
+    )
+}
```

Comment:
This is really one type and one component, with some very minor variation. We can leave it to refactor later, but it would be nice to do it now.

Reply:
Good idea. I have combined them.